src/extension/importer/core/src/main/java/org/geoserver/importer/ImportContext.java

```diff
@@ -278,12 +279,48 @@ public void updated() {
         updateState();
     }
 
-    public void delete() throws IOException {
-        if (data != null) {
-            data.cleanup();
+    public boolean delete() throws IOException {
+        if (canBeRemoved(this)) {
+            for(ImportTask task : this.getTasks()) {
+                if(task.getData() != null) {
+                    if (canBePruned(task)) {
+                        task.getData().cleanup();
+                        return true;
+                    }
+                }
+            }
+        }
+        
+        return false;
+    }
+
+    protected boolean canBeRemoved(ImportContext context) {
+        if (context.getState() == ImportContext.State.COMPLETE || 
+                context.getState() == ImportContext.State.INIT_ERROR)
+            return true;
+        else if (context.getState() == ImportContext.State.PENDING) {
+            for(ImportTask task : context.getTasks()) {
+                if (task.getState() == ImportTask.State.ERROR || task.getError() != null)
```

Comment:
Even an errored task can be updated and re-run, imho the files should not be removed. 
However, this might mean some files are not going to to cleaned up in case of shutdown without cleaning the completed and errored imports. Hum... there could be a shutdown hook that cleans the errored ones in case the memory storage is used, maybe? A non clean shutdown will still leave stuff around of course. Thinking out loud, maybe also add a marker file that indicates incomplete imports? (could be used for pending ones too, but not sure)